runtime/sema/check_binary_expression.go

```diff
@@ -293,18 +297,29 @@ func (checker *Checker) checkBinaryExpressionArithmeticOrNonEqualityComparisonOr
 		)
 	}
 
-	switch operationKind {
-	case BinaryOperationKindArithmetic,
-		BinaryOperationKindBitwise:
-
-		return leftType
+	return leftType
+}
 
-	case BinaryOperationKindNonEqualityComparison:
-		return BoolType
+func (checker *Checker) checkBinaryExpressionNonEquality(
+	expression *ast.BinaryExpression,
+	operation ast.Operation,
+	leftType, rightType Type,
+	anyInvalid bool,
+) (resultType Type) {
+	resultType = BoolType
 
-	default:
-		panic(errors.NewUnreachableError())
+	if !(leftType.Equal(rightType) && leftType.IsComparable() && rightType.IsComparable()) && !anyInvalid {
```

Comment:
Does the requirement that `leftType.Equal(rightType)` here not prevent users from comparing, say, `Int16` values with `Int8` values? This seems like a valuable ability to have.

Reply:
Yes, comparisons, just like e.g arithmetic and bitwise operations do not implicitly convert value types, as the rules are often surprising and lead to bugs.

For example, in C/C++, when comparing signed with unsigned, it is important the signed value is positive to avoid unexpected conversions and results.

Reply:
Yes, there are conversion functions to convert between numeric types